src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/step/rollup/WaitForRollupCompletionStep.kt

```diff
@@ -0,0 +1,132 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.step.rollup
+
+import com.amazon.opendistroforelasticsearch.indexmanagement.elasticapi.suspendUntil
+import com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.model.ManagedIndexMetaData
+import com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.model.managedindexmetadata.StepMetaData
+import com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.step.Step
+import com.amazon.opendistroforelasticsearch.indexmanagement.rollup.action.explain.ExplainRollupAction
+import com.amazon.opendistroforelasticsearch.indexmanagement.rollup.action.explain.ExplainRollupRequest
+import com.amazon.opendistroforelasticsearch.indexmanagement.rollup.action.explain.ExplainRollupResponse
+import com.amazon.opendistroforelasticsearch.indexmanagement.rollup.model.RollupMetadata
+import org.apache.logging.log4j.LogManager
+import org.elasticsearch.client.Client
+import org.elasticsearch.cluster.service.ClusterService
+import java.lang.Exception
+
+class WaitForRollupCompletionStep(
+    val clusterService: ClusterService,
+    val client: Client,
+    managedIndexMetaData: ManagedIndexMetaData
+) : Step(name, managedIndexMetaData) {
+
+    private var stepStatus = StepStatus.STARTING
+    private var info: Map<String, Any>? = null
+    private val logger = LogManager.getLogger(javaClass)
+    private var hasRollupFailed: Boolean? = null
+
+    override fun isIdempotent() = true
+
+    override suspend fun execute(): WaitForRollupCompletionStep {
+        val rollupJobId = managedIndexMetaData.actionMetaData?.actionProperties?.rollupId
+
+        if (rollupJobId == null) {
+            logger.error("No rollup job id passed down")
+            stepStatus = StepStatus.FAILED
+        } else {
+            val explainRollupRequest = ExplainRollupRequest(listOf(rollupJobId))
+            try {
+                val response: ExplainRollupResponse = client.suspendUntil { execute(ExplainRollupAction.INSTANCE, explainRollupRequest, it) }
+                logger.info("Received the status for jobs [${response.getIdsToExplain().keys}]")
+
+                if (response.getIdsToExplain()[rollupJobId]?.metadata?.status == null) {
+                    logger.warn("Job $rollupJobId has not started yet")
+                    stepStatus = StepStatus.CONDITION_NOT_MET
+                    info = mapOf("message" to getJobProcessingMessage(rollupJobId, indexName))
+                } else {
+                    logger.info("Received metadata associated with $rollupJobId")
+                    processRollupMetadataStatus(rollupJobId, response.getIdsToExplain().getValue(rollupJobId)!!.metadata!!)
+                }
+            } catch (e: Exception) {
```

Comment:
Refer to some of the other steps, but you need to handle RemoteTransportExceptions and unwrap the exception.

Reply:
updated